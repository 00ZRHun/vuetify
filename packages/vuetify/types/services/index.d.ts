// Types
import Vue from 'vue'
import { Framework } from 'vuetify'

export interface VuetifyServiceContract {
  framework: Framework
  init: (root: Vue, ssrContext?: object) => void
}

export interface VuetifyService {
  property: string
  new (options?: any): VuetifyServiceContract
}
