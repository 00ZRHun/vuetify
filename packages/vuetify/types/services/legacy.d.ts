export interface Legacy {
  grid: boolean
}
