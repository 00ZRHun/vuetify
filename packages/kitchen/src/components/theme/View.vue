<template>
  <v-content class="pa-1">
    <v-layout row wrap>
      <v-flex sm6 xs12>
        <v-layout column>
          <v-flex px-5 pt-5 pb-0>
            <v-flex
              v-for=" (color, tkey) in themeSelects"
              :key="tkey"
              pa-0
            >
              <v-text-field
                v-model="$vuetify.theme.currentTheme[tkey]"
                solo
                :color="$vuetify.theme[tkey]"
                :label="tkey"
                class="colorField"
              >
                <v-icon
                  slot="prepend-inner"
                  :color="$vuetify.theme.currentTheme[tkey]"
                >
                  mdi-format-color-fill
                </v-icon>
                <v-menu
                  slot="append-outer"
                  transition="slide-y-transition"
                  bottom
                >
                  <v-card
                    slot="activator"
                    :color="$vuetify.theme.currentTheme[tkey]"
                    class="pa-2 d-flex"
                    tile
                    height="48"
                    max-width="48"
                    width="48"
                  />
                  <v-item-group
                    v-model="themeSelects[tkey]"
                    class="hiddenScroll"
                  >
                    <v-item
                      v-for="palette in palettes"
                      :key="palette"
                    >
                      <v-hover slot-scope="{ active, toggle }">
                        <v-card
                          slot-scope="{ hover }"
                          :color="palette"
                          :style="getStyle(active, hover)"
                          :value="palette"
                          :class="`elevation-${active || hover ? '12' : '0'}`"
                          class="pa-2 d-flex"
                          tile
                          height="48"
                          max-width="48"
                          width="48"
                          @click.native="toggle(); setThemeColor(tkey);"
                        />
                      </v-hover>
                    </v-item>
                  </v-item-group>
                </v-menu>
              </v-text-field>
            </v-flex>
          </v-flex>
        </v-layout>
      </v-flex>
      <v-flex sm6 xs12 pa-0>
        <v-layout>
          <v-flex pa-5>
            <v-card
              class="elevation-16 mx-auto"
              color="secondary"
              dark
              height="500"
              max-width="500"
              tile
            >
              <v-system-bar
                class="primary darken-2"
                status
              />
              <v-toolbar
                color="primary"
                flat
                extended
              >
                <v-btn icon>
                  <v-icon>mdi-arrow-left</v-icon>
                </v-btn>
                <v-spacer />
                <v-btn icon>
                  <v-icon>mdi-dots-vertical</v-icon>
                </v-btn>
                <v-toolbar-title
                  slot="extension"
                  class="white--text"
                >
                  Title
                </v-toolbar-title>
                <v-btn
                  color="accent"
                  absolute
                  bottom
                  fab
                  right
                >
                  <v-icon>mdi-share-variant</v-icon>
                </v-btn>
              </v-toolbar>
              <v-layout row wrap class="pt-5">
                <v-flex xs6 px-3>
                  <v-alert type="success" dense />
                </v-flex>
                <v-flex xs6 px-3>
                  <v-alert type="info" dense />
                </v-flex>
                <v-flex xs6 px-3>
                  <v-alert type="warning" dense />
                </v-flex>
                <v-flex xs6 px-3>
                  <v-alert type="error" dense />
                </v-flex>
              </v-layout>
            </v-card>
          </v-flex>
        </v-layout>
      </v-flex>
    </v-layout>
  </v-content>
</template>

<script>
  // Utilities
  import colors from 'vuetify/es5/util/colors'
  import { kebabCase, camelCase } from 'lodash'

  export default {
    components: {
      // ExportModal: () => import('./ExportModal')
    },

    data: () => ({
      colors,
      selections: '',
      exportModal: false,
      themeSelects: {
        primary: '',
        secondary: '',
        accent: '',
        success: '',
        info: '',
        warning: '',
        error: ''
      }
    }),

    computed: {
      palettes () {
        const keys = Object.keys(this.colors)
        return keys.map(kebabCase).slice(0, keys.length - 2)
      }
    },

    methods: {
      getStyle (active, hover) {
        return {
          transform: active || hover ? 'scale(.9, .85)' : 'none'
        }
      },
      setThemeColor (themeProp) {
        const selected = camelCase(this.palettes[this.themeSelects[themeProp]])
        const theme = (this.$vuetify.theme.dark) ? 'dark' : 'light'
        console.log(themeProp, this.themeSelects[themeProp], selected)
        this.$vuetify.theme.setTheme(theme, { [themeProp]: this.colors[selected].base })
      }
    }
  }
</script>
<style>
  .hiddenScroll {
    max-height: 300px;
    overflow-y: scroll;
    background: #fff;
    scrollbar-width: none; /* Firefox */
    -ms-overflow-style: none;  /* IE 10+ */
  }
  .hiddenScroll::-webkit-scrollbar { /* WebKit */
      width: 0px;
  }
  .colorField > .v-input__append-outer {
    margin: 0px !important;
  }
</style>
