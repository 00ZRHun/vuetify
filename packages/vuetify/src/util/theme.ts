import { colorToInt, intToHex, colorToHex, RGB } from './colorUtils'
import * as sRGB from './color/transformSRGB'
import * as LAB from './color/transformCIELAB'
import { VuetifyTheme } from 'vuetify/types'
import colors from './colors'

interface ParsedThemeItem {
  base: string
  lighten5: string
  lighten4: string
  lighten3: string
  lighten2: string
  lighten1: string
  darken1: string
  darken2: string
  darken3: string
  darken4: string

  [name: string]: string
}

interface ParsedTheme {
  [name: string]: ParsedThemeItem
}

export function parse (theme: VuetifyTheme | Record<string, number | string>, isItem = false): ParsedTheme {
  const colors = Object.keys(theme)
  const parsedTheme: any = {}

  for (let i = 0; i < colors.length; ++i) {
    const name = colors[i]
    const value = theme[name]

    if (isItem) {
      if (name === 'base' || name.startsWith('lighten') || name.startsWith('darken')) {
        parsedTheme[name] = colorToHex(value)
      }
    } else if (typeof value === 'object') {
      parsedTheme[name] = parse(value, true)
    } else {
      parsedTheme[name] = genVariations(name, colorToInt(value))
    }
  }

  return parsedTheme
}

/**
 * Generate the CSS for a base color (.primary)
 */
const genBaseColor = (name: string, value: string): string => {
  return `
.${name} {
  background-color: ${value} !important;
  border-color: ${value} !important;
}
.${name}--text {
  color: ${value} !important;
  caret-color: ${value} !important;
}`
}

/**
 * Generate the CSS for a variant color (.primary.darken-2)
 */
const genVariantColor = (name: string, variant: string, value: string): string => {
  const [type, n] = variant.split(/(\d)/, 2)
  return `
.${name}.${type}-${n} {
  background-color: ${value} !important;
  border-color: ${value} !important;
}
.${name}--text.text--${type}-${n} {
  color: ${value} !important;
  caret-color: ${value} !important;
}`
}

const genColorVariableName = (name: string, variant = 'base'): string => `--v-${name}-${variant}`

const genColorVariable = (name: string, variant = 'base'): string => `var(${genColorVariableName(name, variant)})`

export function genStyles (theme: ParsedTheme, cssVar = false): string {
  const colors = Object.keys(theme)

  if (!colors.length) return ''

  let variablesCss = ''
  let css = ''

  const aColor = cssVar ? genColorVariable('primary') : theme.primary.base
  css += `a { color: ${aColor}; }`

  for (let i = 0; i < colors.length; ++i) {
    const name = colors[i]
    const value = theme[name]

    if (typeof value !== 'object') continue

    css += genBaseColor(name, cssVar ? genColorVariable(name) : value.base)
    cssVar && (variablesCss += `  ${genColorVariableName(name)}: ${value.base};\n`)

    const variants = Object.keys(value)
    for (let i = 0; i < variants.length; ++i) {
      const variant = variants[i]
      const variantValue = value[variant]
      if (variant === 'base') continue

      css += genVariantColor(name, variant, cssVar ? genColorVariable(name, variant) : variantValue)
      cssVar && (variablesCss += `  ${genColorVariableName(name, variant)}: ${variantValue};\n`)
    }
  }

  if (cssVar) {
    variablesCss = `:root {\n${variablesCss}}\n\n`
  }

  return variablesCss + css
}

export function genVariations (name: string, value: RGB): Record<string, string> {
  const values: Record<string, string> = {
    base: intToHex(value)
  }

  for (let i = 5; i > 0; --i) {
    values[`lighten${i}`] = intToHex(lighten(value, i))
  }

  for (let i = 1; i <= 4; ++i) {
    values[`darken${i}`] = intToHex(darken(value, i))
  }

  return values
}

export function lighten (value: RGB, amount: number): RGB {
  const lab = LAB.fromXYZ(sRGB.toXYZ(value))
  lab[0] = lab[0] + amount * 10
  return sRGB.fromXYZ(LAB.toXYZ(lab))
}

export function darken (value: RGB, amount: number): RGB {
  const lab = LAB.fromXYZ(sRGB.toXYZ(value))
  lab[0] = lab[0] - amount * 10
  return sRGB.fromXYZ(LAB.toXYZ(lab))
}

export function extractCssColor (color: string, theme: VuetifyTheme): string {
  const parts = color.trim().match(/^([^\s]+)(\s+(lighten|darken)-?([1-5]))?$/) || [null, null, null, null]

  const [, colorName, , variant, num] = parts
  const variantName = variant ? `${variant}${num}` : 'base'

  if (!colorName) {
    throw new TypeError(`Invalid color: ${color}`)
  }

  if (colorName in colors) {
    const colorDef = (colors as any)[colorName]

    if (!colorDef[variantName]) {
      throw new TypeError(`Invalid color: ${color}`)
    }

    return colorDef[variantName]
  }

  if (colorName in theme) {
    const themeDef = theme[colorName] as Record<string, string> | string

    if (typeof themeDef === 'object') {
      if (!themeDef[variantName]) {
        throw new TypeError(`Invalid color: ${color}`)
      }

      return themeDef[variantName]
    }

    if (!variant) {
      return themeDef
    }

    if (variant === 'lighten') {
      return intToHex(lighten(colorToInt(themeDef), Number(num)))
    } else if (variant === 'darken') {
      return intToHex(darken(colorToInt(themeDef), Number(num)))
    }

    throw new TypeError(`Invalid color: ${color}`)
  }

  return color
}
