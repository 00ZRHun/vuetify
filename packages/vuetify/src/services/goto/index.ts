// Extensions
import { Service } from '../service'

// Utilities
import * as easingPatterns from './easing-patterns'
import {
  getContainer,
  getOffset,
  waitForReadystate,
} from './util'

// Types
import Vue from 'vue'
import { GoToOptions, VuetifyGoToTarget } from 'vuetify/types/services/goto'

import { VuetifyServiceContract } from 'vuetify/types/services'

export function scrollBehavior (to, from, savedPosition) {
  return waitForReadystate()
    .then(() => {
      const options = {}
      let scrollTo = 0

      if (to.hash) {
        scrollTo = to.hash
      } else if (savedPosition) {
        scrollTo = savedPosition.y
      }

      return new Promise(resolve => {
        setTimeout(() => {
          if (typeof window === 'undefined') {
            return resolve()
          }

          window.requestAnimationFrame(() => {
            try {
              goTo(scrollTo, options)
            } catch (err) {
              console.log(err)
            }

            resolve()
          })
        }, 200)
      })
    })
}

export default function goTo (
  _target: VuetifyGoToTarget,
  _settings: Partial<GoToOptions> = {}
): Promise<number> {
  const options = goTo.options
  const container = getContainer(options.container)

  /* istanbul ignore else */
  if (options.appOffset && goTo.framework.application) {
    const isDrawer = container.classList.contains('v-navigation-drawer')
    const isClipped = container.classList.contains('v-navigation-drawer--clipped')
    const { bar, top } = goTo.framework.application as any

    options.offset += bar
    /* istanbul ignore else */
    if (!isDrawer || isClipped) options.offset += top
  }

  const startTime = performance.now()

  let targetLocation: number
  if (typeof _target === 'number') {
    targetLocation = getOffset(_target) - options.offset!
  } else {
    targetLocation = getOffset(_target) - getOffset(container) - options.offset!
  }

  const startLocation = container.scrollTop
  if (targetLocation === startLocation) return Promise.resolve(targetLocation)

  const ease = typeof options.easing === 'function'
    ? options.easing
    : easingPatterns[options.easing!]
  /* istanbul ignore else */
  if (!ease) throw new TypeError(`Easing function "${options.easing}" not found.`)

  // Cannot be tested properly in jsdom
  // tslint:disable-next-line:promise-must-complete
  /* istanbul ignore next */
  return new Promise(resolve => requestAnimationFrame(function step (currentTime: number) {
    const timeElapsed = currentTime - startTime
    const progress = Math.abs(options.duration ? Math.min(timeElapsed / options.duration, 1) : 1)

    container.scrollTop = Math.floor(startLocation + (targetLocation - startLocation) * ease(progress))

    const clientHeight = container === document.body ? document.documentElement.clientHeight : container.clientHeight
    if (progress === 1 || clientHeight + container.scrollTop === container.scrollHeight) {
      return resolve(targetLocation)
    }

    requestAnimationFrame(step)
  }))
}

goTo.framework = {} as Record<string, VuetifyServiceContract>
goTo.options = {} as GoToOptions
goTo.init = (root: Vue) => {
  const { router } = root.$options as any

  console.log('router', goTo.options)

  if (!router || !goTo.options.scrollBehavior) return

  console.log('made it')

  router.options.scrollBehavior = scrollBehavior
}

export class Goto extends Service {
  public static property = 'goTo'

  public options: GoToOptions = {
    appOffset: true,
    container: typeof document !== 'undefined' ? (
      (document.scrollingElement as HTMLElement | null) ||
      document.body ||
      document.documentElement
    ) : undefined,
    duration: 500,
    easing: 'easeInOutCubic',
    offset: 0,
    scrollBehavior: false,
  }

  constructor (options: Partial<GoToOptions> = {}) {
    super()

    console.log('init options', options)

    goTo.options = Object.assign({}, this.options, options)

    return goTo
  }
}
