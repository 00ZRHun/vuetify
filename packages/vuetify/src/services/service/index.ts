// Contracts
import { VuetifyServiceContract } from 'vuetify/types/services/index'

// Types
import Vue from 'vue'
import { Framework } from 'vuetify'

export class Service implements VuetifyServiceContract {
  framework!: Framework

  init (root: Vue, ssrContext?: object) {}
}
