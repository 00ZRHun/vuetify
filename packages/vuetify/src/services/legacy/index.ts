// Extensions
import { Service } from '../service'

// Types
import { Legacy as ILegacy } from 'vuetify/types/services/legacy'

export class Legacy extends Service implements ILegacy {
  static property = 'legacy'

  grid = false

  constructor (options: Partial<ILegacy> = {}) {
    super()

    return {
      grid: this.grid,
      framework: {},
      init () {},
    }
  }
}
