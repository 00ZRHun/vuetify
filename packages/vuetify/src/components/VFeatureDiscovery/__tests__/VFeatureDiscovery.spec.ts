// Components
import VFeatureDiscovery from '../VFeatureDiscovery'

// Services
import { Theme } from '../../../services/theme'

// Utilities
import {
  mount,
  Wrapper,
  MountOptions,
} from '@vue/test-utils'

// Libraries
import Vue from 'vue'
const { $createElement: h } = new Vue()

describe('FeatureDiscovery.ts', () => {
  type Instance = InstanceType<typeof VFeatureDiscovery>
  let mountFunction: (options?: MountOptions<Instance>) => Wrapper<Instance>

  beforeEach(() => {
    mountFunction = (options = {} as MountOptions<Instance>) => {
      return mount(VFeatureDiscovery, {
        mocks: {
          $vuetify: {
            theme: new Theme(),
          },
        },
        ...options,
      })
    }
  })

  it('should render', () => {
    const wrapper = mountFunction()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render with activator and activate', () => {
    const wrapper = mountFunction({
      scopedSlots: {
        activator: ({ on }) => h('button', {
          attrs: {
            id: 'foo',
          },
          on,
        }),
      },
    })

    expect(wrapper.html()).toMatchSnapshot()

    expect(wrapper.vm.isActive).toBeFalsy()
    wrapper.find('#foo').trigger('click')
    expect(wrapper.vm.isActive).toBeTruthy()
  })

  it('should render flat component', () => {
    const wrapper = mountFunction({
      propsData: {
        flat: true,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render active component', async () => {
    const wrapper = mountFunction()
    wrapper.setData({
      isActive: true,
    })

    await wrapper.vm.$nextTick()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should deactivate', async () => {
    jest.useFakeTimers()

    const wrapper = mountFunction({
      scopedSlots: {
        activator: ({ on }) => h('button', {
          attrs: {
            id: 'foo',
          },
          on,
        }),
      },
    })
    wrapper.setData({
      isActive: true,
    })
    await wrapper.vm.$nextTick()

    wrapper.setData({
      isActive: false,
    })
    await wrapper.vm.$nextTick()

    const activator = wrapper.vm.getActivator()
    expect(activator.style.zIndex).not.toBe('')
    jest.runAllTimers()
    expect(activator.style.zIndex).toBe('')

    jest.useRealTimers()
  })

  it('should check if activator is fixed', () => {
    const activator = document.createElement('div')

    const wrapper = mountFunction({
      methods: {
        getActivator: () => activator,
      },
    })

    wrapper.vm.checkActivatorFixed()
    expect(wrapper.vm.activatorFixed).toBeFalsy()

    activator.style.position = 'fixed'
    wrapper.vm.checkActivatorFixed()
    expect(wrapper.vm.activatorFixed).toBeTruthy()
  })

  it('should render with custom z-index', () => {
    const wrapper = mountFunction({
      propsData: {
        zIndex: 25,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should measure content', () => {
    const wrapper = mountFunction({
      slots: {
        default: `<div>content</div>`,
      },
    })

    wrapper.vm.$refs.content.getBoundingClientRect = () => ({
      top: 1,
      left: 2,
      bottom: 3,
      right: 4,
      width: 5,
      height: 6,
    })
    wrapper.vm.measureContent()
    expect(wrapper.vm.dimensions.content).toMatchSnapshot()
  })

  it('should calculate backdrop size', () => {
    const wrapper = mountFunction({
      slots: {
        default: `<div>content</div>`,
      },
    })

    wrapper.vm.$refs.content.getBoundingClientRect = () => ({
      top: 0,
      left: 0,
      bottom: 0,
      right: 0,
      width: 120,
      height: 100,
    })
    wrapper.vm.measureContent()

    expect(wrapper.vm.measureDesktopBackdrop()).toMatchObject({
      r: 252.82528798916863,
      size: 585.6505759783372,
      x: 20,
      y: 252.03298642599555,
    })
  })

  it('should render component w/o ripple', () => {
    const wrapper = mountFunction({
      propsData: {
        noRipple: true,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should close on esc', () => {
    const wrapper = mountFunction()
    wrapper.setData({
      isActive: true,
    })

    wrapper.trigger('keydown.esc')
    expect(wrapper.vm.isActive).toBeFalsy()
  })

  it('should close on content click', () => {
    const wrapper = mountFunction()
    const content = wrapper.find('.v-feature-discovery__content')

    wrapper.setData({
      isActive: true,
    })
    content.trigger('click')
    expect(wrapper.vm.isActive).toBeFalsy()

    wrapper.setData({
      isActive: true,
    })
    content.element.setAttribute('disabled', 'disabled')
    content.trigger('click')
    content.element.removeAttribute('disabled')
    expect(wrapper.vm.isActive).toBeTruthy()

    wrapper.setData({
      isActive: true,
    })
    wrapper.setProps({
      closeOnContentClick: false,
    })
    content.trigger('click')
    expect(wrapper.vm.isActive).toBeTruthy()
  })

  it('should change color', () => {
    const wrapper = mountFunction({
      propsData: {
        color: 'red darken-2',
        highlightColor: 'warning lighten-1',
        textColor: '#fefefe',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should update dimensions twice on resize', () => {
    jest.useFakeTimers()

    const wrapper = mountFunction()
    wrapper.setData({
      isActive: true,
    })
    jest.spyOn(wrapper.vm, 'updateDimensions')

    wrapper.vm.onResize()
    jest.runAllTimers()
    expect(wrapper.vm.updateDimensions).toHaveBeenCalledTimes(2)

    jest.useRealTimers()
  })

  it('should get left offset', () => {
    const wrapper = mountFunction()

    ;(window as any).pageXOffset = 10
    expect(wrapper.vm.getOffsetLeft()).toBe(10)

    ;(window as any).pageXOffset = 0
    document.documentElement.scrollLeft = 10
    expect(wrapper.vm.getOffsetLeft()).toBe(10)

    wrapper.vm.hasWindow = false
    expect(wrapper.vm.getOffsetLeft()).toBe(0)
  })

  it('should get top offset', () => {
    const wrapper = mountFunction()

    ;(window as any).pageYOffset = 10
    expect(wrapper.vm.getOffsetTop()).toBe(10)

    ;(window as any).pageYOffset = 0
    document.documentElement.scrollTop = 10
    expect(wrapper.vm.getOffsetTop()).toBe(10)

    wrapper.vm.hasWindow = false
    expect(wrapper.vm.getOffsetTop()).toBe(0)
  })

  describe('sneakPeek', () => {
    const cbs: FrameRequestCallback[] = []
    jest.spyOn(window, 'requestAnimationFrame').mockImplementation(cb => cbs.push(cb))
    const frame = () => cbs.forEach(cb => cb(performance.now()))

    it('should work', () => {
      const wrapper = mountFunction({
        slots: {
          default: `<div>content</div>`,
        },
      })
      const el = wrapper.vm.$refs.content
      el.style.display = 'none'
      const fn = jest.fn(() => expect(el.style.display).toBe('inline-block'))

      wrapper.vm.sneakPeek(fn)

      expect(requestAnimationFrame).toHaveBeenCalled()
      expect(fn).toHaveBeenCalledTimes(0)

      frame()
      expect(fn).toHaveBeenCalledTimes(1)

      expect(el.style.display).toBe('none')
    })

    it('should call callback directly if display != none', () => {
      const wrapper = mountFunction({
        slots: {
          default: `<div>content</div>`,
        },
      })
      const el = wrapper.vm.$refs.content
      el.style.display = 'block'
      const fn = jest.fn(() => expect(el.style.display).toBe('block'))

      wrapper.vm.sneakPeek(fn)

      expect(requestAnimationFrame).toHaveBeenCalled()
      expect(fn).toHaveBeenCalledTimes(0)

      frame()
      expect(fn).toHaveBeenCalledTimes(1)

      expect(el.style.display).toBe('block')
    })

    it('should call callback directly if no content', () => {
      const wrapper = mountFunction({
        methods: {
          genWrap: () => h('div'),
        },
      })

      const el = wrapper.vm.$refs.content
      expect(el).toBeUndefined()

      const fn = jest.fn()

      wrapper.vm.sneakPeek(fn)

      expect(requestAnimationFrame).toHaveBeenCalled()
      expect(fn).toHaveBeenCalledTimes(0)

      frame()
      expect(fn).toHaveBeenCalledTimes(1)
    })
  })
})
