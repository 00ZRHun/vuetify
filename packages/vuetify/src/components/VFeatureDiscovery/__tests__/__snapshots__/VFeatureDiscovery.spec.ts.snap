// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatureDiscovery.ts should change color 1`] = `
<div class="v-feature-discovery">
  <div class="v-feature-discovery__content theme--light"
       style="z-index: 0; color: rgb(254, 254, 254); caret-color: #fefefe; display: none;"
  >
    <div class="v-feature-discovery__backdrop red darken-2"
         style="top: -142.44998779892558px; left: -142.44998779892558px; height: 284.89997559785115px; width: 284.89997559785115px; transform-origin: calc(50% - 0px) calc(50% - 0px);"
    >
    </div>
    <div aria-hidden="true"
         class="v-feature-discovery__highlight warning lighten-1 red--text text--darken-2"
         style="top: -44px; left: -44px; height: 88px; width: 88px;"
    >
    </div>
    <div class="v-feature-discovery__wrap"
         style="top: -64px; left: -40px; height: auto; width: 360px; padding: 0px 40px;"
    >
    </div>
  </div>
</div>
`;

exports[`FeatureDiscovery.ts should measure content 1`] = `
Object {
  "bottom": 3,
  "height": 6,
  "left": 2,
  "right": 4,
  "top": 1,
  "width": 5,
  "x": 5,
  "y": 4,
}
`;

exports[`FeatureDiscovery.ts should render 1`] = `
<div class="v-feature-discovery">
  <div class="v-feature-discovery__content theme--light white--text"
       style="z-index: 0; display: none;"
  >
    <div class="v-feature-discovery__backdrop primary"
         style="top: -142.44998779892558px; left: -142.44998779892558px; height: 284.89997559785115px; width: 284.89997559785115px; transform-origin: calc(50% - 0px) calc(50% - 0px);"
    >
    </div>
    <div aria-hidden="true"
         class="v-feature-discovery__highlight white primary--text"
         style="top: -44px; left: -44px; height: 88px; width: 88px;"
    >
    </div>
    <div class="v-feature-discovery__wrap"
         style="top: -64px; left: -40px; height: auto; width: 360px; padding: 0px 40px;"
    >
    </div>
  </div>
</div>
`;

exports[`FeatureDiscovery.ts should render active component 1`] = `
<div class="v-feature-discovery">
  <div class="v-feature-discovery__content theme--light v-feature-discovery__content--active white--text"
       style="z-index: 8; display: none; z-index: 8;"
  >
    <div class="v-feature-discovery__backdrop primary"
         style="top: -142.44998779892558px; left: -142.44998779892558px; height: 284.89997559785115px; width: 284.89997559785115px; transform-origin: calc(50% - 0px) calc(50% - 0px);"
    >
    </div>
    <div aria-hidden="true"
         class="v-feature-discovery__highlight white primary--text"
         style="top: -44px; left: -44px; height: 88px; width: 88px;"
    >
    </div>
    <div class="v-feature-discovery__wrap"
         style="top: -64px; left: -40px; height: auto; width: 360px; padding: 0px 40px;"
    >
    </div>
  </div>
</div>
`;

exports[`FeatureDiscovery.ts should render component w/o ripple 1`] = `
<div class="v-feature-discovery">
  <div class="v-feature-discovery__content theme--light v-feature-discovery__content--no-ripple white--text"
       style="z-index: 0; display: none;"
  >
    <div class="v-feature-discovery__backdrop primary"
         style="top: -142.44998779892558px; left: -142.44998779892558px; height: 284.89997559785115px; width: 284.89997559785115px; transform-origin: calc(50% - 0px) calc(50% - 0px);"
    >
    </div>
    <div aria-hidden="true"
         class="v-feature-discovery__highlight white primary--text"
         style="top: -44px; left: -44px; height: 88px; width: 88px;"
    >
    </div>
    <div class="v-feature-discovery__wrap"
         style="top: -64px; left: -40px; height: auto; width: 360px; padding: 0px 40px;"
    >
    </div>
  </div>
</div>
`;

exports[`FeatureDiscovery.ts should render flat component 1`] = `
<div class="v-feature-discovery">
  <div class="v-feature-discovery__content theme--light v-feature-discovery__content--flat white--text"
       style="z-index: 0; display: none;"
  >
    <div class="v-feature-discovery__backdrop primary"
         style="top: -142.44998779892558px; left: -142.44998779892558px; height: 284.89997559785115px; width: 284.89997559785115px; transform-origin: calc(50% - 0px) calc(50% - 0px);"
    >
    </div>
    <div aria-hidden="true"
         class="v-feature-discovery__highlight white primary--text"
         style="top: -44px; left: -44px; height: 88px; width: 88px;"
    >
    </div>
    <div class="v-feature-discovery__wrap"
         style="top: -64px; left: -40px; height: auto; width: 360px; padding: 0px 40px;"
    >
    </div>
  </div>
</div>
`;

exports[`FeatureDiscovery.ts should render with activator and activate 1`] = `
<div class="v-feature-discovery v-feature-discovery--inline">
  <button id="foo">
  </button>
  <div class="v-feature-discovery__content theme--light white--text"
       style="z-index: 0; display: none;"
  >
    <div class="v-feature-discovery__backdrop primary"
         style="top: -142.44998779892558px; left: -142.44998779892558px; height: 284.89997559785115px; width: 284.89997559785115px; transform-origin: calc(50% - 0px) calc(50% - 0px);"
    >
    </div>
    <div aria-hidden="true"
         class="v-feature-discovery__highlight white primary--text"
         style="top: -44px; left: -44px; height: 88px; width: 88px;"
    >
    </div>
    <div class="v-feature-discovery__wrap"
         style="top: -64px; left: -40px; height: auto; width: 360px; padding: 0px 40px;"
    >
    </div>
  </div>
</div>
`;

exports[`FeatureDiscovery.ts should render with custom z-index 1`] = `
<div class="v-feature-discovery">
  <div class="v-feature-discovery__content theme--light white--text"
       style="z-index: 25; display: none;"
  >
    <div class="v-feature-discovery__backdrop primary"
         style="top: -142.44998779892558px; left: -142.44998779892558px; height: 284.89997559785115px; width: 284.89997559785115px; transform-origin: calc(50% - 0px) calc(50% - 0px);"
    >
    </div>
    <div aria-hidden="true"
         class="v-feature-discovery__highlight white primary--text"
         style="top: -44px; left: -44px; height: 88px; width: 88px;"
    >
    </div>
    <div class="v-feature-discovery__wrap"
         style="top: -64px; left: -40px; height: auto; width: 360px; padding: 0px 40px;"
    >
    </div>
  </div>
</div>
`;
